parts/xmem_nozzle.go

```diff
@@ -1994,11 +1913,26 @@ func (xmem *XmemNozzle) uncompressBody(req *base.WrappedMCRequest) error {
 	return nil
 }
 func (xmem *XmemNozzle) updateSystemXattrForTarget(wrappedReq *base.WrappedMCRequest, lookup *base.SubdocLookupResponse, setOpt SetMetaXattrOptions) (err error) {
-	if !setOpt.sendHlv && !setOpt.preserveSync {
+	var needToUpdateSysXattrs bool
+	if wrappedReq.Req.DataType&mcc.XattrDataType != 0 {
+		// in this case, we may need to update HLV if source already has it, even if mobile/HLV are all off.
+		needToUpdateSysXattrs = true
+	} else if setOpt.preserveSync {
+		needToUpdateSysXattrs = true
+	} else if xmem.source_cr_mode == base.CRMode_Custom {
+		needToUpdateSysXattrs = true
+	} else if setOpt.sendHlv {
+		maxCas := xmem.config.vbMaxCas[wrappedReq.Req.VBucket]
+		if wrappedReq.Req.Cas >= maxCas {
+			needToUpdateSysXattrs = true
+		}
+	}
+
+	if !needToUpdateSysXattrs {
 		return
 	}
-	if err := xmem.uncompressBody(wrappedReq); err != nil {
-		return err
+	if err = xmem.uncompressBody(wrappedReq); err != nil {
```

Comment:
One thing I realized is that we are uncompressing Body here, but then are we re-compressing the body later? 
If not, we should recompress

Reply:
We call snappy.Encode in preprocessMCRequest()

Reply:
In `updateSystemXattrForTarget` there are 4 cases where decompression will occur:
1. Xattr exists
2. preserveSync
3. CCR mode
4. sendHLV is set

Recompression will happen if:
1. SendHLV is set
2. PreserveSync is set

We are missing Xattr recompression and CCR mode

Can you:
1. Consolidate the logics that is needed for decompression into a function `needToDecompress`
2. and use the same logic as part of the `preprocessMCRequest`

Reply:
We decompress in a few places so it is hard to get the logic right. I will add a field NeedToRecompress in WrappedRequest. We will try to compress for mutations that was compressed.